### PR TITLE
Rewrite to track autoloading and de-register/re-register types appropriately.

### PR DESCRIPTION
**Pull request by pnomolos**
As mentioned in my README changes, I'm sure there are some (or a lot) edge cases that I don't have specs for, but this is working for me in development.

TODO: Update Readme to indicate it should only be included in the `development` group.
TODO: Add a configuration switch to disable/enable dry-types re-registration.
TODO: Include a credit to https://github.com/amhol for the initial thought of tracking at the registration point.

### CODE REVIEW

File: lib/dry/types/rails/registry.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+module Dry
+  module Types
+    module Rails
+      class Registry < ::Dry::Container::Registry
+        def call(container, key, item, options)
+          key = key.to_s.dup.freeze
+          @_mutex.synchronize do
+            container.delete(key) if container.key?(key)
```
Comment (AMHOL) [inline lib/dry/types/rails/registry.rb]: Don't think this line is necessary, can just overwrite with the line below as far as I can tell?

Comment (pnomolos) [inline lib/dry/types/rails/registry.rb]: @AMHOL Just removed the whole custom registry as it wasn't needed (forgot I had left it in) :)

**Comment by jeromegn** [comment]
Wow. That seems great. I'll be able to review that tomorrow.

**Comment by pnomolos** [comment]
@jeromegn Cool.  I'll be actively using this for the foreseeable future so will hopefully be able to keep contributing :)

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -2,11 +2,21 @@
 require 'bundler/setup'
 
 require 'combustion'
+require 'active_support/dependencies'
 
-Combustion.initialize! :action_controller, :action_view, :sprockets
+Combustion.initialize! :action_controller, :action_view, :sprockets do
+  config.eager_load = false
+  config.cache_classes = false
```
Comment (jeromegn) [inline spec/spec_helper.rb:9]: Is this something a Rails app in production would also need to set?

**Comment by jeromegn** [comment]
I had this locally, working on another project. It seemed to work pretty well, but I'm just starting this project and have not tries all the cases.

``` ruby
module Dry
  module Types
    module Rails
      class Railtie < ::Rails::Railtie
        config.to_prepare do
          if Dry::Types.instance_variable_defined? :@container
            Dry::Types.remove_instance_variable :@container
            # HACK: suppress warnings while loading Dry::Types again.
            old_verbose, $VERBOSE = $VERBOSE, nil
            load "dry/types/core.rb"
            $VERBOSE = old_verbose
          end
        end
      end
    end
  end
end
```

What are the advantages of your implementation?

**Comment by pnomolos** [comment]
I'm not at my desk yet this morning, but doing that will remove any types
that weren't autoloaded (and won't re-add them).

**Comment by pnomolos** [comment]
**Edited and re-added this**

I'm not at my desk yet this morning, but doing that will remove any types that weren't autoloaded (and won't re-add them).

To be clearer - it will remove any non-core types that weren't autoloaded (perhaps from another gem).

That solution also needs to have 

``` ruby
Dry::Types::remove_instance_variable :@type_map
```

Added, or the type map will retain references to old types (ex. if you're using a subclass of Dry::Types::Struct).

**Comment by jeromegn** [comment]
Hacky, but LGTM :)

**Comment by jeromegn** [comment]
Pushed to rubygems as 0.3.0

**Pull request merged**